File: gmao-frontend/src/types/purchaseOrder.ts
export type PurchaseOrderStatus =
  | "ALL"
  | "DRAFT"
  | "CONFIRMED"
  | "IN_PROGRESS"
  | "DONE"
  | "CANCELLED"
  | "ARCHIVED";

export type PurchaseOrderLineType = "SPARE_PART" | "FREE_TEXT";

export interface PurchaseOrderLine {
  id: string;
  type: PurchaseOrderLineType;
  sparePartId: number | null;
  sparePartName: string;
  description: string;
  quantity: number;
  unitPrice: number;
  currency: string;
}

export interface PurchaseOrder {
  id: string;
  reference: string;
  supplierId: number | null;
  supplierName: string | null;
  expectedDeliveryDate: string | null;
  notes: string;
  status: Exclude<PurchaseOrderStatus, "ALL">;
  lines: PurchaseOrderLine[];
  createdAt: string;
  updatedAt: string;
}

export interface PurchaseOrderRequest {
  reference: string;
  supplierId: number | null;
  supplierName: string | null;
  expectedDeliveryDate: string | null;
  notes: string;
  lines: PurchaseOrderLine[];
}

File: gmao-backend/src/main/java/gmao_backend/purchaseorder/PurchaseOrder.java
package com.gmao.gmao_backend.purchaseorder;

import jakarta.persistence.CascadeType;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.OneToMany;
import jakarta.persistence.OrderBy;
import jakarta.persistence.PrePersist;
import jakarta.persistence.PreUpdate;
import jakarta.persistence.Table;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "purchase_orders")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class PurchaseOrder {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false, length = 100)
    private String reference;

    @Column(name = "supplier_id")
    private Long supplierId;

    @Column(name = "supplier_name")
    private String supplierName;

    @Column(name = "expected_delivery_date")
    private LocalDate expectedDeliveryDate;

    @Column(columnDefinition = "TEXT")
    private String notes;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 30)
    private PurchaseOrderStatus status;

    @OneToMany(
            mappedBy = "purchaseOrder",
            cascade = CascadeType.ALL,
            orphanRemoval = true,
            fetch = FetchType.LAZY
    )
    @OrderBy("position ASC")
    @Builder.Default
    private List<PurchaseOrderLine> lines = new ArrayList<>();

    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @Column(name = "updated_at", nullable = false)
    private LocalDateTime updatedAt;

    @PrePersist
    void onCreate() {
        LocalDateTime now = LocalDateTime.now();

        if (status == null) {
            status = PurchaseOrderStatus.DRAFT;
        }

        createdAt = now;
        updatedAt = now;
    }

    @PreUpdate
    void onUpdate() {
        updatedAt = LocalDateTime.now();
    }

    public void replaceLines(List<PurchaseOrderLine> newLines) {
        lines.clear();

        for (int index = 0; index < newLines.size(); index += 1) {
            PurchaseOrderLine line = newLines.get(index);
            line.setPurchaseOrder(this);
            line.setPosition(index);
            lines.add(line);
        }
    }
}
